pytorch_lightning/metrics/functional/object_detection.py

```diff
@@ -0,0 +1,125 @@
+
+import torch
+from torch import Tensor
+
+try:
+    from torchvision.ops import box_iou
+except ModuleNotFoundError:  # pragma: no-cover
+    box_iou = None
+
+
+def mean_average_precision(
+    pred_image_indices: Tensor,
+    pred_probs: Tensor,
+    pred_labels: Tensor,
+    pred_bboxes: Tensor,
+    target_image_indices: Tensor,
+    target_labels: Tensor,
+    target_bboxes: Tensor,
+    iou_threshold: float,
+    ap_calculation: str
+) -> Tensor:
+    """
+    Compute mean average precision for object detection task
+
+    Args:
+        pred_image_indices: an (N,)-shaped Tensor of image indices of the predictions
+        pred_probs: an (N,)-shaped Tensor of probabilities of the predictions
+        pred_labels: an (N,)-shaped Tensor of predicted labels
+        pred_bboxes: an (N, 4)-shaped Tensor of predicted bounding boxes
+        target_image_indices: an (M,)-shaped Tensor of image indices of the groudn truths
+        target_labels: an (M,)-shaped Tensor of ground truth labels
+        target_bboxes: an (M, 4)-shaped Tensor of ground truth bounding boxes
+        iou_threshold: threshold for IoU score for determining true positive and
+                       false positive predictions.
+        ap_calculation: method to calculate the average precision of the precision-recall curve
+
+            - ``'step'``: calculate the step function integral, the same way as
+            :func:`~pytorch_lightning.metrics.functional.average_precision.average_precision`
+            - ``'VOC2007'``: calculate the 11-point sampling of interpolation of the precision recall curve
+            - ``'VOC2010'``: calculate the step function integral of the interpolated precision recall curve
+            - ``'COCO'``: calculate the 101-point sampling of the interpolated precision recall curve
+
+    Returns:
+        mean of the average precisions of all classes in object detection task.
+
+    References:
+        - host.robots.ox.ac.uk/pascal/VOC/
+        - https://ccc.inaoep.mx/~villasen/bib/AN%20OVERVIEW%20OF%20EVALUATION%20METHODS%20IN%20TREC%20AD%20HOC%20IR%20AND%20TREC%20QA.pdf#page=15
+        - https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py
+    """
+    if box_iou is None:
+        raise ImportError('`mean_average_precision` metric requires `torchvision`, which is not installed. '
+                          ' install it with `pip install torchvision`.')
+    classes = torch.cat([pred_labels, target_labels]).unique()
+    average_precisions = torch.zeros(len(classes))
+    for class_idx, c in enumerate(classes):
+        # Descending indices w.r.t. class probability for class c
+        desc_indices = torch.argsort(pred_probs, descending=True)[pred_labels == c]
+        # No predictions for this class so average precision is 0
+        if len(desc_indices) == 0:
+            continue
+        targets_per_images = Counter([idx.item() for idx in target_image_indices[target_labels == c]])
```

Comment:
Can we somehow avoid the idx.item() here in favor of tensor-based-indexing? This might cause a GPU/TPU sync and blocking memory transfer which could be slow.

an alternative could be something like this:

```python
curr_target_image_indices = target_image_indices[target_labels == c]
targets_per_image = torch.zeros((curr_target_image_indices.unique().numel(),), device=torch.device)
for idx in targets_per_image.size(0):
    targets_per_image[idx] += (curr_target_image_indices == idx).long().sum()
```
I am not sure if this is better, but at least it does not do any obvious synchronizations.